dbms/src/Core/Spiller.cpp

```diff
@@ -46,35 +47,43 @@ SpilledFile::~SpilledFile()
     }
 }
 
-Spiller::Spiller(const String & id_, bool is_input_sorted_, size_t partition_num_, const String & spill_dir_, const Block & input_schema_, const LoggerPtr & logger_)
-    : id(id_)
+Spiller::Spiller(const SpillConfig & config_, bool is_input_sorted_, size_t partition_num_, const Block & input_schema_, const LoggerPtr & logger_)
+    : config(config_)
     , is_input_sorted(is_input_sorted_)
     , partition_num(partition_num_)
-    , spill_dir(spill_dir_)
     , input_schema(input_schema_)
     , logger(logger_)
 {
-    if (spill_dir.at(spill_dir.size() - 1) != Poco::Path::separator())
-    {
-        spill_dir += Poco::Path::separator();
-    }
     for (size_t i = 0; i < partition_num; ++i)
         spilled_files.push_back(std::make_unique<SpilledFiles>());
+    Poco::File spill_dir(config.spill_dir);
+    if (!spill_dir.exists())
+    {
+        /// just for test, usually the tmp path should be created when server starting
+        spill_dir.createDirectories();
+    }
+    else
+    {
+        if (!spill_dir.isDirectory())
+            throw Exception(fmt::format("Spilled dir {} is a file", spill_dir.path()));
+    }
 }
 
-bool Spiller::spillBlocks(const Blocks & blocks, size_t partition_id)
+void Spiller::spillBlocks(const Blocks & blocks, size_t partition_id)
 {
-    RUNTIME_CHECK_MSG(partition_id < partition_num, "{}: partition id {} exceeds partition num {}.", id, partition_id, partition_num);
-    RUNTIME_CHECK_MSG(spill_finished == false, "{}: spill after the spiller is finished.", id);
-    /// todo append to existing file
+    RUNTIME_CHECK_MSG(partition_id < partition_num, "{}: partition id {} exceeds partition num {}.", config.spill_id, partition_id, partition_num);
+    RUNTIME_CHECK_MSG(spill_finished == false, "{}: spill after the spiller is finished.", config.spill_id);
+    /// todo 1. append to existing file, 2. check disk usage
     if (unlikely(blocks.empty()))
-        return true;
+        return;
+    has_spilled_data = true;
     auto spilled_file_name = nextSpillFileName(partition_id);
+    LOG_INFO(logger, "Spilling part of data into temporary file {}", spilled_file_name);
     try
     {
         auto spilled_file = std::make_unique<SpilledFile>(spilled_file_name);
         RUNTIME_CHECK_MSG(!spilled_file->exists(), "Duplicated spilled file: {}, should not happens", spilled_file_name);
-        WriteBufferFromFile file_buf(spilled_file_name);
+        WriteBufferFromFileProvider file_buf(config.file_provider, spilled_file_name, EncryptionPath(spilled_file_name, ""));
```

Comment:
Please use `FileProvider::deleteRegularFile(spilled_file_path, EncryptionPath(spilled_file_path, "")` to delete the spilled files in the destructor of `SpilledFile`.

Reply:
OK, updated the code.